src/main/java/com/minecolonies/coremod/colony/CitizenData.java

```diff
@@ -464,19 +491,119 @@ private String generateName(@NotNull final Random rand)
+     *
+     * @param rand Random object.
+     */
+    public void generateName(@NotNull final Random rand, final String firstParentName, final String secondParentName)
+    {
+        String nameA = firstParentName;
+        String nameB = secondParentName;
+
+        String citizenName;
+        final String firstName;
+        final String middleInitial;
+        final String lastName;
+
+        if (firstParentName == null || firstParentName.isEmpty())
+        {
+            nameA = generateName(rand, rand.nextBoolean(), colony);
+        }
+
+        if (secondParentName == null || secondParentName.isEmpty())
+        {
+            nameB = generateName(rand, rand.nextBoolean(), colony);
+        }
+
+        final String[] firstParentNameSplit = nameA.split(" ");
+        final String[] secondParentNameSplit = nameB.split(" ");
+
+        int lastNameIndex = 1;
+        if (MineColonies.getConfig().getServer().useEasternNameOrder.get())
+        {
+            lastNameIndex = 0;
+        }
+        else if (MineColonies.getConfig().getServer().useMiddleInitial.get())
+        {
+            lastNameIndex = 2;
+        }
+
+        if (random.nextBoolean())
+        {
+            middleInitial = firstParentNameSplit[lastNameIndex].substring(0, 1);
+            lastName = secondParentNameSplit[lastNameIndex];
+        }
+        else
+        {
+            middleInitial = secondParentNameSplit[lastNameIndex].substring(0, 1);
+            lastName = firstParentNameSplit[lastNameIndex];
+        }
+
+        if (female)
+        {
+            firstName = getRandomElement(rand, MineColonies.getConfig().getServer().femaleFirstNames.get().toArray(new String[0]));
+        }
+        else
+        {
+            firstName = getRandomElement(rand, MineColonies.getConfig().getServer().maleFirstNames.get().toArray(new String[0]));
+        }
+
+        if(MineColonies.getConfig().getServer().useEasternNameOrder.get())
+        {
+            //For now, don't include middle names, as their rules (and presence) vary heavily by region.
+            citizenName = String.format("%s %s", lastName, firstName);
+        }
+        else
+        {
+            if (MineColonies.getConfig().getServer().useMiddleInitial.get())
+            {
+                citizenName = String.format("%s %s. %s", firstName, middleInitial, lastName);
+            }
+            else
+            {
+                citizenName = String.format("%s %s", firstName, lastName);
+            }
+        }
+
+        // Check whether there's already a citizen with this name
+        for (final ICitizenData citizen : this.getColony().getCitizenManager().getCitizens())
+        {
+            if (citizen != null && citizen.getName().equals(citizenName))
+            {
+                // Oops - recurse this function and try again
+                generateName(rand, firstParentName, secondParentName);
+                break;
```

Comment:
Doesn't this code generate a non-duplicated name and then overwrite that name with the name that has a duplicate?

Reply:
yes, this has to be a return